Tests/ParsingTests/OptionallyTests.swift

```diff
@@ -14,4 +14,18 @@ final class OptionalTests: XCTestCase {
     XCTAssertNoDifference(.none, Optionally { Bool.parser() }.parse(&input))
     XCTAssertNoDifference("Hello, world!", Substring(input))
   }
+    
+  func testBacktracking() {
+    let parser = Parse {
+      "Hello,"
+      Optionally {
+        " "
+        Bool.parser()
+      }
+      " world!"
+    }
+    
+    XCTAssertNoDifference(.some(true), try parser.parse("Hello, true world!"))
+    XCTAssertNoDifference(.none, try parser.parse("Hello, world!"))
+  }
```

Comment:
I've updated the `testBacktracking` test case to illustrate where `Optionally` will not parse as expected if there is no backtracking.

The first assertion of `"Hello, true world!"` will pass with or without backtracking, because the `Optionally` doesn't fail and correctly consumes the `" true"` string (including the leading space).

The second assertion of `"Hello, world!"` will fail without backtracking, because the space will be consumed inside the `Optionally` and will not be available for the `" world!"` parser.

Reply:
Nice, that does make a lot of sense. Only other option I can think of to consider is whether we introduce an explicit `.backtracking` operator to opt into backtracking, and continue leaving all other backtracking to just `OneOf`. But the fact that `Optionally` is just sugar for `OneOf` does make me think it should backtrack.

So, I do think this is a good change, but will wait to merge it in for a few days because @stephencelis is traveling right now. Also I think we should make this change on main and then bring into the printer branch separately since this doesn't depend on anything from printing. Could you rebase it when you have a chance?

Reply:
Yeah, I think it produces more unexpected results if there is no backtracking. I was scratching my head for a bit initially when it popped up in my parser.

My changes do tweak the `print` method also, and throws an error if the print fails, although that is perhaps a different issue.

Reply:
In general, I think that any `Parser` which catches errors and may continue on without throwing its own error should backtrack. Otherwise you will end up with surprising results for parsers further down the line.

Reply:
Yeah I think we can bring the print changes back once the parse stuff is in main. And then we can worry about tests and documentation for that behavior at that time too.

In fact, I suppose docs could be updated to mention the backtracking behavior, but I can take a stab at that once it merges if you don't have time.

Reply:
> In general, I think that any `Parser` which catches errors and may continue on without throwing its own error should backtrack. Otherwise you will end up with surprising results for parsers further down the line.

Yeah that's interesting. We should keep that in mind for future documentation. We're working on a [DocC](https://github.com/pointfreeco/swift-parsing/compare/docc) overhaul so that we have better places to call out info like this.